Fix duplicate database change event send on Library.add

valrus opened the pull request:
## Description

Fixes #5560. Also a couple other incidental changes / improvements:
* Add `EventType` that holds the actual string literals used for event sending. With type checking, this can prevent subtle bugs resulting from misspelled event names.
* Allow `is_hidden` to take `str`, and convert those `str` to `Path`. Without this fix, `test_hidden.py` was failing for me due to `path.lstat()` — `test_osx_hidden` passes a `str`, and `str` doesn't have a `lstat` method.

## To Do

- [x] ~Documentation.~
- [x] Changelog.
- [x] Tests.

JOJ0 reviewed (commented):
This looks like a useful bugfix. It looks reasonable to me but I'm a little out of my depth here and since it's rather small to review would ask for your opinions @snejus and @wisp3rwind if you have a minute or two?

JOJ0 commented on beets/util/hidden.py:
```diff
@@ -23,13 +23,15 @@
     if isinstance(path, bytes):
         path = Path(os.fsdecode(path))
+    elif isinstance(path, str):
+        path = Path(path)
```
Especially about this change I'm not sure. `is_hidden` is only called here throughout the codebase:

https://github.com/beetbox/beets/blob/5d010e95ec8940d59aac9f9c0dd987a77f617dd8/beets/util/__init__.py#L248

It is made sure that paths are wrapped using the `bytestring_path` util. So shouldn't we better fix `test_is_hidden` to wrap using `bytestring_path` rather than accepting `str` here?

valrus replied:
Yes absolutely. As I just re-discovered, there's a [section](https://github.com/beetbox/beets/blob/master/CONTRIBUTING.rst#handling-paths) in CONTRIBUTING.rst about it. I made that change.

JOJ0 reviewed (approved):
Thanks for the updates! Please for documentation's sake update the description with the final state. Other than that I think this is good to go :-)

wisp3rwind commented:
> This looks like a useful bugfix. It looks reasonable to me but I'm a little out of my depth here and since it's rather small to review would ask for your opinions @snejus and @wisp3rwind if you have a minute or two?

I'm also not very familiar with the intricacies of `Library`. At a quick glance, this change seems correct (In the sense that we don't needlessly duplicate the event; I don't think there is a real guarantee that it is generally impossible for it to be sent in close succession with no actual changes.)

Regarding the `test_hidden` change, the current (added `bytestring_path`) version looks good to me.

Small nitpicks:
- I'd prefer the new file not to be named `types.py`: That's very generic, we already have two or so of these in other places, and I generally find anything called just `{t,T}ype{,s}` to easily cause confusion regarding what it might refer to. Maybe it could be `events.py`? Specifically, in the context of https://github.com/beetbox/beets/pull/5701, I've been wondering whether it might eventually be possible to provide (optionally) typed events (i.e. in some way specify the arguments that are passed along). That would probably require some additional infrastructure, which would fit well into an `events` module.
- `DatabaseChangeTestBase` -> `DatabaseChangeTest`. Maybe, this should actually be part of `test_library.py`? In our current tests, we tend to collect tests for a each module in a single file, rather than have one per test case.
- In general, I'd prefer this to be 3 separate PRs, focusing each on only one of the (independent!) parts of this PR. That tends to simplify review, and allows quick merging of uncontentious smaller pieces. Right now, I'm fine with keeping this PR as it is, and merge things in one go.

valrus commented:
Thanks for the review! I'll address those nitpicks when I have a chance.

I agree smaller PRs would have been better. This was a case of setting out to fix one thing and stumbling into some other improvements along the way, and by the time I'd done what I actually intended in the first place it was tangled up with the other stuff. I would like to be more disciplined about this and sometimes I am, but it's easier when I'm more familiar with the codebase.

valrus commented:
Fixed:
* `types.py` renamed to `event_types.py`. I opted for this instead of `events.py` because it does only have types for now, and I've found that having modules that contain only types decreases the likelihood of circular import snafus because types are unlikely to depend on other things outside the stdlib. It can always be renamed to `events.py` later if its surface area expands.
* Moved the new test into `beets.test.test_library::AddTest` and deleted its old location.

JOJ0 commented:
except conflict resolving, this is good to go. thanks for the input @wisp3rwind , I agree that splitting concerns would be the best approach. this time it's certainly ok like it is.

JOJ0 commented:
And appologies for not reacting earlier, which would have prevented a lot of conflicts in the first place. I'm moving my home to a new flat and don't have much FOSS time is the excuse ;-)

JOJ0 commented:
> > can you rebase to fewer commits? removing the ones that were not part of the final solution. For example the kwarg you removed in a later commit again? sorry more work but I stumbled across them again while rereading commits one by one. Thanks!!! 🙏
> 
> 
> 
> Alternatively, you could squash-merge the PR :)

I had that idea but wasn't sure about it because there are at least two different things in this PR. And in case bugs happen later pinpointing to/reverting one commit/thing would be easier. But actually the test hidden fix is not a big deal and most probably correct. So yes I agree let's finally merge it  and we all can move on :-)
